### DISCUSSION TITLE
Extend bundle

### DISCUSSION THREAD

**Discussion by YusufSuleman** [bevyengine/bevy#9232] [Q&A]
How do I extend a bundle in bevy, like there is an Object bundle containing, things like Position, Physics data, etc. And then there is a Player bundle, an enemy bundle, platform bundle, etc.
How would I do this? A simple google search leads me no where

Edit: I would need to extend those bundles further. like a harder enemy bundle, rotating platform bundle.

**Comment by Zoomulator** [comment]
There are examples of this in the docs :)
https://docs.rs/bevy/latest/bevy/ecs/bundle/trait.Bundle.html#implementors-1

Hope it helps!

**Reply by YusufSuleman** [reply]
No, not a bundle inside a bundle, extending it.

**Reply by Zoomulator** [reply]
That's how you extend/combine different aspects tho.

The Player bundle would include the Object bundle if it requires those fields. It all gets flattened out in the ECS, and each system adds behaviors based on components, not the specific bundles.

**Reply by YusufSuleman** [reply]
If I try to spawn, it requires the object struct field, which you say should be flattened,
```rust
    #[bundle(ignore)]
    pub object: Object,
```

**Reply by Zoomulator** [reply]
I don't think you want the `#[bundle(ignore)]` attribute there :)

Just add the field, no attribute is required:
```
#[derive(Component)]
struct Health(f32);

#[derive(Component)]
struct Enemy;

#[derive(Bundle)]
struct CharacterBaseBundle {
  spatial: SpatialBundle,
  health: Health,
}

#[derive(Bundle)]
struct EnemyBundle {
  character: CharacterBaseBundle,
  enemy_tag: Enemy,
}

fn my_system(mut commands: Commands) {
  commands.spawn(EnemyBundle);
}
```

**Reply by YusufSuleman** [reply]
No, what if I want to extend an object that extends an object that extends an object that extends an object... Then I would have a lot of nesting when spawning these.

Edit: I want clean code that is easy to handle, when I need to add new things, I know the cost of dirty code.

**Reply by Zoomulator** [reply]
In my previous example: EnemyBundle extends CharacterBaseBundle that extends SpatialBundle. The resulting enemy will have all the components from those bundles.

Not sure, but maybe you need to utilize defaults to get what you need in terms of initialization.

Note that the new bundles now derive `Default` as well.

```
#[derive(Component)]
struct Health(f32);

#[derive(Component)]
struct Enemy;

#[derive(Bundle, Default)]
struct CharacterBaseBundle {
  spatial: SpatialBundle,
  health: Health,
}

#[derive(Bundle, Default)]
struct EnemyBundle {
  character: CharacterBaseBundle,
  enemy_tag: Enemy,
}

fn my_system(mut commands: Commands) {
  commands.spawn(EnemyBundle {
    character: CharacterBaseBundle {
       health: Health(5.0),
       ..default()
    },
    ..default()
  });
}
```

Also it's probably a good idea to make things data-driven. For instance, most "enemies" will likely have the same component layout, but their behaviors are impacted by what data those components carry. Maybe it's preferable to create a factory function that assembles these entities rather than providing unique bundles for them.

**Reply by Zoomulator** [reply]
Worth knowing that you can also add components dynamically to your entities if need be:
```
fn my_system(mut commands: Commands) {
  let id = commands.spawn(MyStartingBundle);
  commands.entity(id).insert_component(Health(5.0));
}
```

**Reply by YusufSuleman** [reply]
So something like this:
```rust
#[derive(Bundle)]
pub struct Object {
    pub transform: Transform
}

#[derive(Bundle)]
pub struct BasicObject {
    pub object: Object,
    pub sprite: SpriteBundle
}

pub fn basic_object_factory() -> (BasicObject) {
    (BasicObject {
        object: Object { transform: Transform {..Default::default()} },
        sprite: SpriteBundle {..Default::default()},
    })
}
```

I am making a 2D game not 3D, I know it seems useless to have Object, but it creates a bit cleaner code and writing Transform everywhere would look dirty or would this make it even dirtier? I am getting a bit confused now?

**Reply by Zoomulator** [reply]
SpriteBundle already includes the Transform component actually: https://docs.rs/bevy/latest/bevy/prelude/struct.SpriteBundle.html

Also if you find the nesting a bit cumbersome, just assign the default to a local variable before spawning it and do your changes.

```rust
fn spawning_system(mut commands: Commands) {
  let mut obj = BasicObject::default();
  obj.sprite.transform = Transform::from_translation(Vec3::new(1.0, 2.0, 3.0));
  let id = commands.spawn(obj);

  if this_thing_needs_more_stuff() {
    commands.entity(id).insert(AnotherBundle::default()); // insert takes bundles, just like spawn!
  }
}
```

**Reply by YusufSuleman** [reply]
Should I just use SpriteBundle, or just include all the sprite things in it without the Transform and GlobalTransform?

**Reply by Zoomulator** [reply]
I'd just use SpriteBundle  ¯\_(ツ)_/¯
